proms-resources/frontend/src/main/frontend/src/QuestionnaireSet.jsx

```diff
@@ -0,0 +1,407 @@
+
+  // Data already associated with the subject
+  const [ subjectData, setSubjectData ] = useState();
+  // Did the user make it to the last screen?
+  const [ endReached, setEndReached ] = useState();
+  // Has everything been filled out?
+  const [ isComplete, setComplete ] = useState();
+
+  // Step -1: haven't started yet, welcome screen
+  // Step i = 0 ... # of questionnaires-1 : filling out questionnaire #i
+  // Step n = # of questionnaires: all done, exist screen
+  const [ crtStep, setCrtStep ] = useState(-1);
+  // What questionnaire comes next after the current step
+  const [ nextQuestionnaire, setNextQuestionnaire] = useState(null);
+  // What form we're currently displaying
+  const [ crtFormId, setCrtFormId ] = useState(null);
+  // When something goes wrong:
+  const [ error, setError ] = useState("");
+  // Screen layout props
+  const [ screenType, setScreenType ] = useState ();
+
+  const classes = useStyles();
+
+  const globalLoginDisplay = useContext(GlobalLoginContext);
+
+  // Find the data already associated with the subject
+  useEffect(() => {
+    if (!questionnaires) return;
+    loadExistingData();
+  }, [questionnaires]);
+
+  // Determine the screen type (and style) based on the step number
+  useEffect(() => {
+    setScreenType(crtStep >= 0 && crtStep < questionnaireIds.length ? "survey" : "screen");
+  }, [crtStep]);
+
+  // Reset the crtFormId when returning to the welcome screen
+  useEffect(() => {
+    crtStep == -1 && setCrtFormId(null);
+  }, [crtStep]);
+
+  // Determine the next questionnaire that needs to be filled out
+  useEffect(() => {
+    if (!questionnaires || !subjectData) return;
+    // Find the next unfilled questionnaire, if any:
+    let nextStep = findNextStep(crtStep);
+    setNextQuestionnaire(nextStep < questionnaireIds?.length ? questionnaires[questionnaireIds?.[nextStep]] : null);
+  }, [crtStep, questionnaires, subjectData]);
+
+  // If we're back to the start because the user was directed to add missing answers,
+  // dont't show the welcome screen and skip to the next step without them pressing start
+  useEffect(() => {
+    if (crtStep == -1 && endReached) {
+      nextQuestionnaire && launchNextForm();
+    }
+  }, [crtStep, endReached, nextQuestionnaire]);
+
+  // When the uuid of the next form is set, move to the next step to launch the form
+  useEffect(() => {
+    crtFormId && nextStep();
+  }, [crtFormId]);
+
+  // At the last step, determine if the questionnaires have been ceompleted
+  useEffect(() => {
+    if (!questionnaireIds || crtStep < questionnaireIds.length) return;
+    setEndReached(true);
+    loadExistingData();
+  }, [crtStep, questionnaireIds]);
+
+  // Determine if all surveys have been filled out
+  useEffect(() => {
+    if (!subjectData || !questionnaires) return;
+    setComplete(Object.keys(subjectData || {}).filter(q => isFormComplete(q)).length == questionnaireIds.length);
+  }, [subjectData, questionnaires]);
```

Comment:
It's not clear to me why the conditions are on `questionnaires` but the code itself uses `questionnaireIds`. Shouldn't `questionaireIds` be used throughout?

```suggestion
    if (!subjectData || !questionnaireIds) return;
    setComplete(Object.keys(subjectData || {}).filter(q => isFormComplete(q)).length == questionnaireIds.length);
  }, [subjectData, questionnaireIds]);
```